### PR TITLE
Pass rule details in evaluation context

### PR DESCRIPTION
**Pull request by dannykopping**
Grafana Loki uses Prometheus' rule evaluation engine, and we need to log the rule name and labels during evaluation.

This implementation mimicks the existing mechanism to expose the rule group and filename via the context in `rules/manager.go`:

```go
	ctx = promql.NewOriginContext(ctx, map[string]interface{}{
		"ruleGroup": map[string]string{
			"file": g.File(),
			"name": g.Name(),
		},
	})
```

It diverges slightly to make the context data type-safe.
I'm happy to either amend the existing implementation to follow the same pattern, or lose the type-safety and accept a `map[string]string` as above.

### CODE REVIEW

**Comment by dannykopping** [comment]
@roidelapluie I see you worked on the existing implementation I referenced in the description; would you mind taking a look at this?

File: rules/origin.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// NewRuleDetail creates a RuleDetail from a given Rule.
+func NewRuleDetail(r Rule) RuleDetail {
+	var kind string
+	switch r.(type) {
+	case *AlertingRule:
+		kind = KindAlerting
+	case *RecordingRule:
+		kind = KindRecording
+	default:
+		kind = "unknown"
```
Comment (roidelapluie) [inline rules/origin.go:46]: We should probably panic here.

Comment (dannykopping) [reply rules/origin.go:46]: Done, and test added :+1:

**Comment by dannykopping** [comment]
@roidelapluie can you please have another look?

**Review by roidelapluie — APPROVED** [review]

**Comment by roidelapluie** [comment]
Thanks!

**Comment by roidelapluie** [comment]
Thanks!

**Pull request merged**